src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncPivotBlockSelector.scala

```diff
@@ -0,0 +1,144 @@
+
+class FastSyncPivotBlockSelector(
+    val etcPeerManager: ActorRef,
+    val peerEventBus: ActorRef,
+    val syncConfig: SyncConfig,
+    val scheduler: Scheduler
+) extends Actor
+    with ActorLogging
+    with PeerListSupport
+    with BlacklistSupport {
+
+  import FastSyncPivotBlockSelector._
+  import syncConfig._
+
+  val fastSync: ActorRef = context.parent
+
+  def handleCommonMessages: Receive = handlePeerListMessages orElse handleBlacklistMessages
+
+  override def receive: Receive = idle
+
+  def idle: Receive = handleCommonMessages orElse { case ChoosePivotBlock =>
+    val peersUsedToChooseTarget = peersToDownloadFrom.collect { case (peer, PeerInfo(_, _, true, maxBlockNumber, _)) =>
+      (peer, maxBlockNumber)
+    }
+
+    if (peersUsedToChooseTarget.size >= minPeersToChoosePivotBlock) {
+      tryChooseTargetBlock(peersUsedToChooseTarget)
+    } else {
+      log.info(
+        "Cannot pick pivot block. Need at least {} peers, but there are only {} available at the moment. Retrying in {}",
+        minPeersToChoosePivotBlock,
+        peersUsedToChooseTarget.size,
+        startRetryInterval
+      )
+      scheduleRetry(startRetryInterval)
+      context become idle
+    }
+  }
+
+  def tryChooseTargetBlock(peersWithBestBlockNumbers: Map[Peer, BigInt]): Unit = {
+    val peersSortedByBestNumber = peersWithBestBlockNumbers.toList.sortBy(-_._2)
+    val bestPeerBestBlockNumber = peersSortedByBestNumber.head._2
+    val expectedPivotBlock = (bestPeerBestBlockNumber - syncConfig.pivotBlockOffset).max(0)
+    val peersToAsk = peersSortedByBestNumber.takeWhile(_._2 >= expectedPivotBlock).map(_._1)
+
+    log.info(
+      "Trying to choose fast sync pivot block using {} peers. The best block is {}. Ask {} peers for block nr {}",
+      peersWithBestBlockNumbers.size,
+      bestPeerBestBlockNumber,
+      peersToAsk.size,
+      expectedPivotBlock
+    )
+
+    peersToAsk.foreach { peer =>
+      peerEventBus ! Subscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peer.id)))
+      etcPeerManager ! EtcPeerManagerActor.SendMessage(
+        GetBlockHeaders(Left(expectedPivotBlock), 1, 0, reverse = false),
+        peer.id
+      )
+    }
+
+    val timeout = scheduler.scheduleOnce(peerResponseTimeout, self, PivotBlockTimeout)
+    context become waitingForPivotBlock(peersToAsk.map(_.id).toSet, expectedPivotBlock, timeout, Map.empty)
+  }
+
+  def waitingForPivotBlock(
+      peersToAsk: Set[PeerId],
+      targetBlockNumber: BigInt,
+      timeout: Cancellable,
+      headers: Map[BlockHeader, Int]
+  ): Receive =
+    handleCommonMessages orElse {
+      case MessageFromPeer(blockHeaders: BlockHeaders, peerId) =>
+        peerEventBus ! Unsubscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peerId)))
+        val updatedPeersToAsk = peersToAsk - peerId
+        val targetBlockHeaderOpt = blockHeaders.headers.find(header => header.number == targetBlockNumber)
```

Comment:
maybe we should validate that we received exactly one header with required number ?